mbedtls/src/ssl/context.rs

```diff
@@ -6,42 +6,39 @@
  * option. This file may not be copied, modified, or distributed except
  * according to those terms. */
 
+
+use core::any::Any;
+use core::result::Result as StdResult;
 #[cfg(not(feature = "std"))]
-use core_io::{self as io, Read, Write};
+use core_io::{Read, Write, Result as IoResult, Error as IoError, ErrorKind as IoErrorKind};
+#[cfg(feature = "std")]
+use std::io::{Read, Write, Result as IoResult, Error as IoError, ErrorKind as IoErrorKind};
 #[cfg(feature = "std")]
-use std::io::{self, Read, Write};
+use std::sync::Arc;
 
 use mbedtls_sys::types::raw_types::{c_int, c_uchar, c_void};
 use mbedtls_sys::types::size_t;
 use mbedtls_sys::*;
 
-use crate::error::{Error, IntoResult, Result};
-use core::result::Result as StdResult;
+use crate::alloc::{List as MbedtlsList};
+#[cfg(not(feature = "std"))]
+use crate::alloc_prelude::*;
+use crate::error::{Error, Result, IntoResult};
+use crate::pk::Pk;
 use crate::private::UnsafeFrom;
-use crate::ssl::config::{AuthMode, Config, Version};
-use crate::x509::{Crl, LinkedCertificate, VerifyError};
+use crate::ssl::config::{Config, Version, AuthMode};
+use crate::x509::{Certificate, Crl, VerifyError};
 
-pub trait IoCallback {
-    unsafe extern "C" fn call_recv(
-        user_data: *mut c_void,
-        data: *mut c_uchar,
-        len: size_t,
-    ) -> c_int;
-    unsafe extern "C" fn call_send(
-        user_data: *mut c_void,
-        data: *const c_uchar,
-        len: size_t,
-    ) -> c_int;
 
+pub trait IoCallback {
+    unsafe extern "C" fn call_recv(user_data: *mut c_void, data: *mut c_uchar, len: size_t) -> c_int where Self: Sized;
+    unsafe extern "C" fn call_send(user_data: *mut c_void, data: *const c_uchar, len: size_t) -> c_int where Self: Sized;
     fn data_ptr(&mut self) -> *mut c_void;
+    fn as_any(&mut self) -> &mut dyn Any;
```

Comment:
That would make it harder to handle the Context<T> types.

I use this trick as I can store the IO as dyn pointer and avoid templates on Context.
```
        io: Option<Box<dyn IoCallback>>,
```

This allows easier handling of Context class as I don't need to know what T is, only the callers care when they call:
```
    pub fn get_mut_io<T: IoCallback + 'static>(&mut self) -> IoResult<&mut T> {
```

Easier handling means easy casting to Context when callback from mbedtls is triggered among other things.

Reply:
Thanks for pointing out we can setup static dispatch for the most part before converting to a trait object. That's good. But instead of adding this method to this trait, I think you should do this:

```
trait IoCallback: Any { ... }

struct Context {
    io: Option<Box<dyn Any>>,
    ...
}

impl Context {
    fn io(&self) -> &dyn Any { ... }
    fn io_mut(&mut self) -> &mut dyn Any { ... }
    // possibly also add:
    fn into_io(self) -> Box<dyn Any> { ... }
}
```